File: Dart/Behavioral/observer.dart
/// Simple class to put a message notification
class Notification {
  Notification({
    required this.message,
  });

  final String message;
}

/// Interface of subscriber class
abstract interface class Subscriber {
  void notify({required Notification notification});
}

/// Concrete implementation of susbscriber.
/// In this case we are using the user name.
class NotificationSubscriber implements Subscriber {
  NotificationSubscriber({required this.name});

  final String name;

  @override
  void notify({required Notification notification}) =>
      print('$name, ${notification.message}');

  @override
  toString() => '$name';
}

/// Class to manage the subscribers and notify them.
class Publisher {
  List<Subscriber> _subscribers = [];

  void registerSubscriber({required Subscriber subscriber}) {
    _subscribers.add(subscriber);
    print('$subscriber added!');
  }

  void removeSubscriber({required Subscriber subscriber}) {
    _subscribers.remove(subscriber);
    print('$subscriber removed!');
  }

  void notifySubscribers({required Notification notification}) => _subscribers
      .forEach((subscriber) => subscriber.notify(notification: notification));
}

void main() {
  final publisher = Publisher();
  final notification = Notification(message: 'you have a new notification!');
  final user1 = NotificationSubscriber(name: 'User1');
  final user2 = NotificationSubscriber(name: 'User2');
  final user3 = NotificationSubscriber(name: 'User3');

  publisher
    ..registerSubscriber(subscriber: user1)
    ..registerSubscriber(subscriber: user2);

  publisher.notifySubscribers(notification: notification);

  publisher
    ..registerSubscriber(subscriber: user3)
    ..removeSubscriber(subscriber: user1);

  publisher.notifySubscribers(notification: notification);

  /// Here is the result:
/*
User1 added!
User2 added!
User1, you have a new notification!
User2, you have a new notification!
User3 added!
User1 removed!
User2, you have a new notification!
User3, you have a new notification!
*/
}

File: Swift/Behavioral/observer.swift
//
//  observer.swift
//
//

import Foundation

// Simple struct to hold a message notification
struct Notification {
    let message: String

    init(message: String) {
        self.message = message
    }
}

// Protocol for subscriber types
protocol Subscriber {
    var id: UUID { get }
    func notify(notification: Notification)
}

// Concrete implementation of Subscriber using the user's name
struct NotificationSubscriber: Subscriber, CustomStringConvertible {
    let id: UUID
    let name: String

    init(name: String) {
        self.id = UUID()
        self.name = name
    }

    func notify(notification: Notification) {
        print("\(name), \(notification.message)")
    }
    
    // Custom description String
    var description: String { name }
}

// Class to manage subscribers and notify them
final class Publisher {
    private var subscribers: [any Subscriber] = []

    func registerSubscriber(subscriber: any Subscriber) {
        subscribers.append(subscriber)
        print("\(subscriber) added!")
    }

    func removeSubscriber(subscriber: any Subscriber) {
        subscribers.removeAll { $0.id == subscriber.id }
        print("\(subscriber) removed!")
    }

    func notifySubscribers(notification: Notification) {
        subscribers.forEach { $0.notify(notification: notification) }
    }
}

// Demo

let publisher = Publisher()
let notification = Notification(message: "you have a new notification!")
let user1 = NotificationSubscriber(name: "User1")
let user2 = NotificationSubscriber(name: "User2")
let user3 = NotificationSubscriber(name: "User3")

publisher.registerSubscriber(subscriber: user1)
publisher.registerSubscriber(subscriber: user2)

publisher.notifySubscribers(notification: notification)

publisher.registerSubscriber(subscriber: user3)
publisher.removeSubscriber(subscriber: user1)

publisher.notifySubscribers(notification: notification)

// Here is the result:

// User1 added!
// User2 added!
// User1, you have a new notification!
// User2, you have a new notification!
// User3 added!
// User1 removed!
// User2, you have a new notification!
// User3, you have a new notification!
